WIP - OpenGraph and link embed fallbacks

notnownikki opened the pull request:
## Description

Adds OpenGraph and plain link fallbacks to oEmbed rendering.

I'd like feedback on what we render for OpenGraph, and the user experience, especially around pasting links that have no embed or OG data.

## Screenshots
Embedding a youtube video, https://www.youtube.com/watch?v=PfKUdmTq2MI

![embed-paste-youtube](https://user-images.githubusercontent.com/3314354/36255572-78577f54-1247-11e8-9492-491e7078b794.gif)

Embedding with OpenGraph, https://reactjs.org/docs/rendering-elements.html

![embed-paste-opengraph](https://user-images.githubusercontent.com/3314354/36255622-a9d22d86-1247-11e8-86de-b88e2c0a8083.gif)

Embedding a site with no embed support and no OpenGraph data, http://notnownikki.com/sameties/

![embed-paste-url](https://user-images.githubusercontent.com/3314354/36255649-be2a51aa-1247-11e8-9313-35518d3d2204.gif)

notnownikki committed:
WIP - OpenGraph and link embed fallbacks

notnownikki commented:
There are issues with this that I know about :)

The format it returns breaks classic editor embeds, and the html it returns for a URL you can't embed doesn't work well for existing URLs in posts that can't be embedded. Please (for now) ignore these issues, as it's a WIP and I'm fixing them.

jasmussen commented:
I love this. Thank you for working on it. 

Pasting youtube works great.

Pasting https://reactjs.org/docs/rendering-elements.html works good.

Pasting http://notnownikki.com/sameties/ works completely fine. 

This post is also supposed to have OG data: https://www.theverge.com/2018/2/6/16977756/black-panther-movie-review-ryan-coogler-marvel-cinematic-universe-chadwick-boseman-michael-b-jordan — though this did not seem to do anything at all.

Opengraph links, opengraph blocks, pasting and fetching data — this is a rabbit hole, and I want to be very clear that even if we just get this one PR in, the whole pasting process will be VASTLY improved, as such I wouldn't find it necessary to make any improvements outside of what's already here. Nice work.

_However_, one of the long term improvements that would be very cool to see, and this is low priority and anyone who has the time available to do this in the future shoudl be free to, would be to go a few steps further in the direction of Facebook and social networks:

- There's an explicit OpenGraph block. It's samesies as the video embeds, except with a different description.
- The link pasted is more designed, like a card, a sans serif font, and a thumbnail.
- We can perhaps even offer inspector options to change the layout of this OpenGraph card — one layout features a big thumbnail up top, another features a small thumbnail on the side, a third features no thumbnail.
- Even posts that do not feature any opengraph data, we crawl out and look for thumbnails and page text, to stitch together our own. 
- Perhaps as an author you can pick the thumbnail between a few options.

I like how Google Plus does this. Try pasting any link in the writing prompt. Here's a post with OG data:

![screen shot 2018-02-19 at 08 48 56](https://user-images.githubusercontent.com/1204802/36367447-faeb60a8-1552-11e8-8090-24e5f088f70c.png)

Here's your videogame!

![screen shot 2018-02-19 at 08 48 05](https://user-images.githubusercontent.com/1204802/36367434-efcde376-1552-11e8-8dfc-e1ade2623c2e.png)

The X button removes the card entirely. The little "back and forward" icon lets you pick a different thumbnail.

notnownikki commented:
@jasmussen thank you for the feedback :)

I'll take a look at the page from the verge, the OG data _is_ there, but something is stopping us from fetching and parsing it.

The next steps here are for us to have a Gutenberg embed controller that extends the existing WordPress one, and adds the needed filters only during the request. Adding them for all requests to the current embed controller breaks the old editor, and that's bad.

Also, would anyone have time to help with getting a design that uses the OG image too? I'm not sure I want to tackle that myself, as we need eyes who know more about how this will appear in published posts, too.

jasmussen commented:
Feel free to just output the OG image and then ping me and I'll put some card layout together :) thank you.

notnownikki commented:
Great! When this is done, it'll extend the oEmbed behaviour to fetch OG content so the same HTML will be output in a published post. The bit that interferes with the old editor is converting it into a link with the title, so that's the filter that will only get applied during a Gutenberg request.

notnownikki commented:
Gah, `WP_oEmbed_Controller` is final, so I can't extend and do stuff with it. For now, I'll just implement the OG fallback, and skip getting the title of the page to set as the link text (as that's the bit I'd need to extend to do cleanly).

notnownikki committed:
Fix

notnownikki committed:
Lint fixes

notnownikki commented:
@jasmussen Branch has been updated, you'll need to pull this down fresh because it's been rebased.

If you paste a URL into an empty paragraph block, you won't see the usual embed UI while it does the embed, so that the UI doesn't jump around.

The OpenGraph fallback is in there for embedding in Gutenberg. This doesn't work in the old editor, it just shows the URL.

But, the OpenGraph code does run on the post itself, and shows on the front end.

We need to sort out the rendering of the image, as the size isn't correct. The whole thing is cached just like the rest of embeds are, so you'll need to disable that while you're changing things. To do that, comment out line 580 in `lib/register.php`. The HTML is generated in `gutenberg_oembed_opengraph` in `lib/register.php`

jasmussen commented:
Really nice work. I'm going to push some visual tweaks to it. 

**Question:** For the use cases where you really DO want to paste just the link, on an empty line, and not have it transform into an OG card, can we do exactly what Google Plus does?

![plus](https://user-images.githubusercontent.com/1204802/36412182-63626ad4-1619-11e8-8f1e-55b842e426bb.gif)

Note how the text of the URL pasted is actually still there. You can then delete this text and keep only the OG card, or delete the OG card and only keep the text/link. 

To convert this to Gutenberg parlance — can we do it so the link pasted remains in a separate paragraph block, and an OG block with the link is inserted below? 

Asking as a question because if this is onerous or a lot of dev work, do let me know and we can look at other options.

Commit:
Add OpenGraph styles

These styles are very basic and minimalist, and they do not allow the card to scale up vertically, which solves a problem with the iframe not resizing either.

jasmussen commented:
I pushed some visuals to the OG card:

![screen shot 2018-02-20 at 08 57 06](https://user-images.githubusercontent.com/1204802/36412851-2d84aaa0-161c-11e8-8cff-47a3cc5f0c12.png)

I know that looks slightly unimpressive, but I think we might as well mimic Facebook style basic link cards for now, then we can consider adding more styles in the future. 

I noticed that if you click the link inside the card, you get a white empty iframe. Perhaps this goes back to the "cover the embed" issue we had a while back — it'd be fine if you couldn't click the link at all in the editor.

notnownikki commented:
> For the use cases where you really DO want to paste just the link, on an empty line, and not have it transform into an OG card, can we do exactly what Google Plus does?

Tricky.

The thing is, if we had a paragraph block with the URL by itself, and an OpenGraph embed block, the paragraph block would still get processed by oEmbed when the post is shown on the front end of the site.

The way to do it would be a custom OpenGraph embed block that had all the UI and options to hide the card, or hide bits of the card, or show/hide the URL.

However, to do that, we've got to transform the embed block into the appropriate embed block based on the embed provider, and we don't currently have the code to do that.

So - not a simple bit of work. But, one we absolutely should do, because we're getting interest in adding Twitter specific options to the Twitter block, and that would be much more useful if we could make sure all twitter URLs ended up in a Twitter embed block.

jasmussen commented:
Gotcha, then don't worry about it. 👍 👍 

This is a far superior experience to what was before, I think it's worth continuing on the path we're on.

notnownikki commented:
I think the next step is to transform blocks into the correct embed blocks based on the embed provider, then we can start getting custom blocks in there. I can see there was work recently to get the provider when we do an embed request, so that's a good start. From there it's a matter of being able to look up block settings based on that provider, and replacing the current block with the right one. The look up is the trickiest part, but we might be able to do something with naming conventions to avoid a separate registry. I'd quite like to tackle that next :)

jasmussen commented:
We may need @youknowriad's input on the following one:

![link embedding](https://user-images.githubusercontent.com/1204802/36418346-e0d587fc-162e-11e8-9fb8-08f2aa204c74.gif)

When I deselect a link pasted on the first line that doesn't have OG data, it looks like the writing prompt still has placeholder text.

notnownikki commented:
Yes, and the cursor isn't placed so that the user can immediately start typing. We definitely need help on how to do that :+1:

notnownikki committed:
Shift the OG styles around

notnownikki commented:
I shifted the OG styles inline, into `gutenberg_oembed_opengraph` and they display OK in the editor, but there seems to be a floating issue when they're displayed on the front end, and the link and description are pushed out of sight. @jasmussen could you take a look please?

jasmussen commented:
I'm not seeing the ported styles anymore — do I have to clear some cache or something to rebuild? I ran npm run build too:

![screen shot 2018-02-20 at 12 15 19](https://user-images.githubusercontent.com/1204802/36421172-c5c786fa-1637-11e8-8f6e-94ebc6d66378.png)

notnownikki commented:
Yes, you can either clear all your transients by running the following SQL:

```DELETE FROM wp_options WHERE option_name LIKE '_transient_%';```

or, comment out lines 532-538 in `lib/register.php`

Commit:
Try to adjust the inline styles.

jasmussen commented:
Okay I think I pushed a fix. Let me know if this works for you!

iseulde commented on lib/register.php:
```diff
@@ -517,3 +517,133 @@ function gutenberg_revisions_restore( $revisions_data ) {
+	$image = empty( $image_match ) ? '' : $image_match[1];
+	preg_match( '/<meta[^>]+property="og:description"[^>]*>/i', $body, $desc_match );
+	$desc = empty( $desc_match ) ? '' : $desc_match[0];
+	preg_match( '/content="([^"]+)"/i', $desc, $desc_match );
```
I know I kind of quickly put it together like this, but could we maybe make a separate function for this? Like `getOGPropertyContent( $body, $property )`?

notnownikki replied:
Good call :) I'd like to have all of those parsed with a single regex if possible, and just return an array of OG data.

iseulde replied:
Sounds good too!

notnownikki commented:
Moving the style to an inline style block is great! Much more readable :)

I'm still seeing some issues with the link and description through. Here are a couple of screenshots, and the HTML that's in the page...

In the editor (all looks great!)

![screenshot from 2018-02-20 12-19-51](https://user-images.githubusercontent.com/3314354/36423689-cd35e58e-1638-11e8-87b6-499563b5b111.png)

In the post, description and link are not visible:

![screenshot from 2018-02-20 12-20-11](https://user-images.githubusercontent.com/3314354/36423706-dbe68ffc-1638-11e8-8f20-79eb6c841284.png)

Here's the HTML copied from the post page:

```
    <div class="entry-content">
        <figure class="wp-block-embed is-type-opengraph">
    <blockquote><img src="https://cdn.vox-cdn.com/thumbor/NOpbKZQQRSXlieJbOJjdYstCWfw=/0x0:2040x1068/fit-in/1200x630/cdn.vox-cdn.com/uploads/chorus_asset/file/10164247/BlackPanther596d2f04d1540_2040.jpg" /><h1><a href="https://www.theverge.com/2018/2/6/16977756/black-panther-movie-review-ryan-coogler-marvel-cinematic-universe-chadwick-boseman-michael-b-jordan">Black Panther review: the grown-up Marvel movie we’ve been waiting for &#8211; The Verge</a></h1><p>Director Ryan Coogler delivers one of the best films in the franchise.</p></blockquote><style type="text/css">
        body.opengraph blockquote,
        .is-type-opengraph blockquote {
            margin: 0;
            border: 1px solid #e1e3e6;
            border-radius: 4px;
            padding: 8px;
            overflow: hidden;
            height: 96px;
            box-sizing: content-box;
            -moz-box-sizing: content-box;
            -webkit-box-sizing: content-box;
        }
        body.opengraph img,
        .is-type-opengraph img {
            width: 96px;
            height: 96px;
            object-fit: cover;
            float: left;
            margin-right: 1em;
        }
        body.opengraph p,
        body.opengraph h1,
        .is-type-opengraph p,
        .is-type-opengraph h1 {
            font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif;
            font-size: 13px;
            margin: 0;
        }
        body.opengraph a {
            color: #007daa;
        }
        </style>
</figure>
    </div>
```

iseulde commented on lib/register.php:
```diff
@@ -517,3 +517,133 @@ function gutenberg_revisions_restore( $revisions_data ) {
+ * @return mixed OpenGraph HTML, or false if there was no OpenGraph data.
+ */
+function gutenberg_oembed_opengraph( $url ) {
+	$cache_key = 'gutenberg_opengraph_' . md5( $url );
```
Is it common to use `md5` for stuff like this? Can it just be the escaped URL?

notnownikki replied:
This is how the transient keys for embeds are generated, bit of a cargo cult copy-and-paste here I'm afraid, because I figured "well, that's how WP does it".

iseulde commented on lib/register.php:
```diff
@@ -517,3 +517,133 @@ function gutenberg_revisions_restore( $revisions_data ) {
+		$url,
+		array(
+			'limit_response_size' => 250000,
+			'user-agent'          => 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:20.0) Gecko/20100101 Firefox/20.0',
```
Why is this needed? If there's a good reason, could you leave a comment?

notnownikki replied:
Without a user agent, sites like The Verge block requests. Will add a comment!

iseulde replied:
Interesting. Would it be possible to just pass the current user agent? Saves some space and also feels less like cheating haha :) Also this would make is so requests come from a variety of UAs, not just a single one which Gould be detected later?

notnownikki replied:
:+1: will do that!

iseulde commented on blocks/library/embed/index.js:
```diff
@@ -156,6 +163,13 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 					</BlockControls>
 				);
 
+				if ( fetching && fromPaste ) {
```
Why do we need to differentiate from normal fetch?

notnownikki replied:
This seems like a good candidate for a comment :)

If I pasted a URL into a paragraph, then I'd see the "Embedding..." UI from the Embed block pop up while the fetch happened. It caused a bit of UI jumping and looked out of place. With this, we see just a paragraph with the link in it while the fetch goes on, and it feels like you don't leave the paragraph block while that's happening.

iseulde replied:
Sure, but don't we then convert to an embed block at the end anyway?

notnownikki replied:
If we can embed the URL, we do. If we can't, it becomes a paragraph block. So with this flow, the user sees "Paragraph block -> pasted URL in a paragrph -> (Embedded content | URL in paragraph block)"

Without it, user sees, "Paragraph block -> Large grey block saying "Embedding" -> (Embedded content | URL in paragraph block)" and the experience doesn't feel as smooth.

iseulde replied:
Yep, this is assuming we don't have the paragraph fallback then.

notnownikki replied:
If we didn't have paragraph fallback, this can go completely. I'd like to tackle that separately though, because if we're going to go the whole Facebook style "guess the preview", then we need to give the user more control over what gets displayed as part of that preview, or allow them to get rid of it completely. That's quite a bit more work, and I think this is a good first step on the way to getting there.

notnownikki replied:
To expand on this, to keep oEmbed happy, pasting a URL that oEmbed cannot handle would result in a paragraph block with the URL in it, and an "Embedded URL Preview" block (or some similar name) which generates a preview separate from the link in the paragraph block. That's so the preview can be optional, and have bits edited and customised, should our guesses get it wrong.

iseulde commented on blocks/library/embed/index.js:
```diff
@@ -123,9 +125,14 @@ function getEmbedBlockSettings( { title, icon, category = 'embed', transforms, k
 
+							// invalid url, and this didn't come from submitting the form
+							if ( 'oembed_invalid_url' === code && ! event ) {
+								this.props.onReplace( createBlock( 'core/paragraph', { content: ( <a href={ url }>{ url }</a> ) } ) );
```
Could we maybe always try to grab something from the page, even if it's just the title/URL/favicon/image/summary of the page? Saw similar suggestions on the PR, but also would make this redundant.

notnownikki replied:
I'd like to do that, and I'd like to look at it in another PR once we have the mechanism working here first. I get the feeling we'd want to give more control over what's presented for non-embeddable URLs, and this would probably go well with a custom "non-embeddable embed" block that would have options for editing the description, etc. There's some other work we need to do to make that happen, so I didn't want to delay this.

iseulde replied:
Ah I see what you mean where you can sometimes choose the image. Maybe for now, instead of the OG title we can just display the content in the title tags (in card form)? Nothing more?

notnownikki replied:
I had a version of this that did that for URLs that didn't have OG data, but it resulted in the oembed proxy endpoint returning data  for every URL, and never saying "I can't embed this" and that caused chaos in the classic editor. That's why I ended up thinking that we need an embed block specifically for this that uses a different API and saves preview HTML as a block that the oEmbed code doesn't touch.

iseulde commented on components/sandbox/index.js:
```diff
@@ -134,7 +134,6 @@ export default class Sandbox extends Component {
 				document.body.style.position = 'absolute';
 				document.body.style.width = '100%';
 				document.body.setAttribute( 'data-resizable-iframe-connected', '' );
-
```
Accidental? :)

notnownikki replied:
`git commit -a --amend`

YOU SAW NOTHING ;)

iseulde commented:
After seeing the inline HTML on the front-end, can't we use the style.css sheet for blocks? Then it is not repeated if there is more than one?

iseulde commented:
Regarding going back to just a plain link... in core we (1) add an undo level so you if you press cmd+z you get the link and (2) ESC also gives you the plain link again, acting exactly the same as the undo level.

notnownikki commented:
@iseulde yeah, there are a couple of issues I need help with when pasting an unsupported link. The undo is one of them, the other is placing the cursor so that the user can paste and carry on typing without having to navigate to the end of the link.

notnownikki committed:
review points

notnownikki committed:
lint

notnownikki commented:
Pushed some fixes up from the reviews.

@jasmussen if you could take a look at the front end rendering? The title and description are pushed out of view, with the default theme at least.

@iseulde I made the OG data parsing a bit better, there's just a regex that extracts it all now, and I added comments to the places where the reasoning needed to be explained. Unsupported URLs get put as links in a paragraph block, and I'd appreciate your input on how to handle undo, and how to put the cursor at the end of the content so the user can just carry on typing.

Once we have this landed, I'll start on the "URL Preview" block to support URLs that can't be handled by oEmbed, and then we should have an experience that matches Facebook/G+ :)

Commit:
Fix some CSS bleed for the OG embed.

jasmussen commented:
Okay I believe the embed issue should be fixed now:

![screen shot 2018-02-21 at 12 30 00](https://user-images.githubusercontent.com/1204802/36477884-0b836e4a-1703-11e8-8164-33ff0a74d520.png)

notnownikki commented:
@jasmussen works for me, and looks great! Thank you!

notnownikki commented:
I'm going to close this PR, but keep the branch around because there are issues with installs that have custom embed providers that don't quite do what oembed does, and don't return the metadata we need completely.

I'll take a different approach to getting the correct embed block type, and implement the OpenGraph and URL Preview blocks too.
